File: frontend/src/lib/validation/content.ts
import { z } from "zod";

// TODO(M4b): generate from backend content.schema.json
const markSchema = z.discriminatedUnion("type", [
  z.object({ type: z.literal("bold") }).strict(),
  z.object({ type: z.literal("italic") }).strict(),
  z.object({ type: z.literal("underline") }).strict(),
  z.object({ type: z.literal("subscript") }).strict(),
  z.object({ type: z.literal("superscript") }).strict(),
  z.object({
    type: z.literal("link"),
    attrs: z.object({
      href: z.string(),
      title: z.string().nullable().optional(),
      target: z.string().nullable().optional(),
      rel: z.string().nullable().optional(),
    }).strict(),
  }).strict(),
]);

const leafMarksSchema = z.union([z.number().min(0), z.string().min(1), z.null()]).optional();

export type ContentNode = {
  type: string;
  attrs?: Record<string, unknown>;
  content?: ContentNode[];
  text?: string;
  marks?: z.infer<typeof markSchema>[];
};
export type QuestionContent = { type: "doc"; marks?: string | number | null; content: ContentNode[] };

const textNode = z.object({ type: z.literal("text"), text: z.string(), marks: z.array(markSchema).optional() }).strict();
const hardBreakNode = z.object({ type: z.literal("hardBreak") }).strict();
const imageNode = z.object({
  type: z.literal("image"),
  attrs: z.object({ assetId: z.string().uuid(), alt: z.string().optional(), widthMm: z.number().positive().optional() }).strict(),
}).strict();
const answerSpaceNode = z.object({
  type: z.literal("answerSpace"),
  attrs: z.union([
    z.object({ lines: z.number().int().min(1) }).strict(),
    z.object({ blankHeightMm: z.number().min(1) }).strict(),
  ]),
}).strict();

const inlineNodeSchema: z.ZodType<ContentNode, ContentNode> = z.lazy(() => z.union([textNode, hardBreakNode]));
const blockNodeSchema: z.ZodType<ContentNode, ContentNode> = z.lazy(() => z.union([
  z.object({ type: z.literal("paragraph"), content: z.array(inlineNodeSchema).optional() }).strict(),
  z.object({ type: z.literal("heading"), attrs: z.object({ level: z.union([z.literal(1), z.literal(2), z.literal(3)]) }).strict(), content: z.array(inlineNodeSchema).optional() }).strict(),
  z.object({ type: z.literal("bulletList"), content: z.array(listItemSchema).min(1) }).strict(),
  z.object({ type: z.literal("orderedList"), content: z.array(listItemSchema).min(1) }).strict(),
  tableSchema,
  imageNode,
  z.object({
    type: z.literal("subQuestion"),
    attrs: z.object({
      label: z.string().min(1).nullable().optional(),
      marks: leafMarksSchema,
    }).strict().optional(),
    content: z.array(blockNodeSchema).min(1),
  }).strict(),
  answerSpaceNode,
]));

const listItemSchema: z.ZodType<ContentNode, ContentNode> = z.lazy(() => z.object({
  type: z.literal("listItem"),
  content: z.tuple([z.object({ type: z.literal("paragraph"), content: z.array(inlineNodeSchema).optional() }).strict()]).rest(blockNodeSchema),
}).strict());
const tableCellSchema: z.ZodType<ContentNode, ContentNode> = z.lazy(() => z.object({ type: z.literal("tableCell"), content: z.array(blockNodeSchema).min(1) }).strict());
const tableRowSchema: z.ZodType<ContentNode, ContentNode> = z.lazy(() => z.object({ type: z.literal("tableRow"), content: z.array(tableCellSchema).min(1) }).strict());
const tableSchema: z.ZodType<ContentNode, ContentNode> = z.lazy(() => z.object({ type: z.literal("table"), content: z.array(tableRowSchema).min(1) }).strict());

export const contentSchema: z.ZodType<QuestionContent, QuestionContent> = z.object({ type: z.literal("doc"), marks: leafMarksSchema, content: z.array(blockNodeSchema) }).strict();

export function isValidContent(json: unknown): json is QuestionContent {
  return contentSchema.safeParse(json).success;
}

export function normalizeContentForWire(content: QuestionContent): QuestionContent {
  const normalize = (node: ContentNode): ContentNode => {
    const attrs = node.attrs ? { ...node.attrs } : undefined;
    if (node.type === "image" && attrs?.alt === null) delete attrs.alt;
    // Strip only null/undefined marks — a real 0 is authoritative and must be kept.
    if (attrs && attrs.marks == null) delete attrs.marks;
    return {
      ...node,
      ...(attrs ? { attrs } : {}),
      ...(node.content ? { content: node.content.map(normalize) } : {}),
    };
  };
  return {
    type: "doc",
    ...(content.marks == null ? {} : { marks: content.marks }),
    content: content.content.map(normalize),
  };
}

export function hasSubQuestions(content: QuestionContent): boolean {
  return content.content.some((n) => n.type === "subQuestion");
}

function numberOf(raw: unknown): number {
  if (raw == null) return 0;
  const n = Number(raw);
  return Number.isFinite(n) && n >= 0 ? n : 0;
}

export function leafMarksTotal(content: QuestionContent): number {
  const sumSubs = (nodes: ContentNode[]): number =>
    nodes
      .filter((n) => n.type === "subQuestion")
      .reduce((acc, sub) => {
        const nested = (sub.content ?? []).some((n) => n.type === "subQuestion");
        return nested ? acc + sumSubs(sub.content ?? []) : acc + numberOf(sub.attrs?.marks);
      }, 0);
  if (hasSubQuestions(content)) return sumSubs(content.content);
  return numberOf(content.marks);
}

File: backend/app/schemas/content.py
"""Strict canonical ProseMirror/TipTap content schema.

Run ``python -m app.schemas.content`` to print the matching JSON Schema.
"""

from __future__ import annotations

import json
from typing import Annotated, Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, TypeAdapter, model_validator


class StrictModel(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)


class EmptyAttrs(StrictModel):
    pass


class LinkAttrs(StrictModel):
    href: str = Field(min_length=1)
    title: str | None = None
    target: Literal["_blank", "_self"] | None = None


class BoldMark(StrictModel):
    type: Literal["bold"]


class ItalicMark(StrictModel):
    type: Literal["italic"]


class UnderlineMark(StrictModel):
    type: Literal["underline"]


class SubscriptMark(StrictModel):
    type: Literal["subscript"]


class SuperscriptMark(StrictModel):
    type: Literal["superscript"]


class LinkMark(StrictModel):
    type: Literal["link"]
    attrs: LinkAttrs


Mark = Annotated[
    BoldMark | ItalicMark | UnderlineMark | SubscriptMark | SuperscriptMark | LinkMark,
    Field(discriminator="type"),
]


class TextNode(StrictModel):
    type: Literal["text"]
    text: str
    marks: list[Mark] = Field(default_factory=list)


class HardBreakNode(StrictModel):
    type: Literal["hardBreak"]


class ImageAttrs(StrictModel):
    asset_id: UUID
    alt: str = ""
    title: str | None = None


class ImageNode(StrictModel):
    type: Literal["image"]
    attrs: ImageAttrs


InlineNode = Annotated[TextNode | HardBreakNode | ImageNode, Field(discriminator="type")]


class ParagraphNode(StrictModel):
    type: Literal["paragraph"]
    content: list[InlineNode] = Field(default_factory=list)


class HeadingAttrs(StrictModel):
    level: Literal[1, 2, 3]


class HeadingNode(StrictModel):
    type: Literal["heading"]
    attrs: HeadingAttrs
    content: list[InlineNode] = Field(default_factory=list)


class AnswerSpaceAttrs(StrictModel):
    lines: int | None = Field(default=None, ge=0)
    blank_height_mm: float | None = Field(default=None, alias="blankHeightMm", ge=0)

    @model_validator(mode="after")
    def exactly_one_dimension(self) -> AnswerSpaceAttrs:
        if (self.lines is None) == (self.blank_height_mm is None):
            raise ValueError("answerSpace requires exactly one of lines or blankHeightMm")
        return self


class AnswerSpaceNode(StrictModel):
    type: Literal["answerSpace"]
    attrs: AnswerSpaceAttrs


class ListItemNode(StrictModel):
    type: Literal["listItem"]
    content: list[BlockNode] = Field(min_length=1)


class BulletListNode(StrictModel):
    type: Literal["bulletList"]
    content: list[ListItemNode] = Field(min_length=1)


class OrderedListAttrs(StrictModel):
    start: int = Field(default=1, ge=1)


class OrderedListNode(StrictModel):
    type: Literal["orderedList"]
    attrs: OrderedListAttrs = Field(default_factory=OrderedListAttrs)
    content: list[ListItemNode] = Field(min_length=1)


class TableCellNode(StrictModel):
    type: Literal["tableCell"]
    content: list[BlockNode] = Field(min_length=1)


class TableRowNode(StrictModel):
    type: Literal["tableRow"]
    content: list[TableCellNode] = Field(min_length=1)


class TableNode(StrictModel):
    type: Literal["table"]
    content: list[TableRowNode] = Field(min_length=1)

    @model_validator(mode="after")
    def rectangular(self) -> TableNode:
        widths = {len(row.content) for row in self.content}
        if len(widths) != 1:
            raise ValueError("table rows must contain the same number of cells")
        return self


class SubQuestionAttrs(StrictModel):
    label: str = Field(min_length=1)


class SubQuestionNode(StrictModel):
    type: Literal["subQuestion"]
    attrs: SubQuestionAttrs
    content: list[BlockNode] = Field(min_length=1)


BlockNode = Annotated[
    ParagraphNode
    | HeadingNode
    | BulletListNode
    | OrderedListNode
    | TableNode
    | ImageNode
    | SubQuestionNode
    | AnswerSpaceNode,
    Field(discriminator="type"),
]


class DocNode(StrictModel):
    type: Literal["doc"] = "doc"
    content: list[BlockNode] = Field(default_factory=list)


DocNode.model_rebuild()
CONTENT_ADAPTER = TypeAdapter(DocNode)


def parse_content(value: object) -> DocNode:
    """Validate untrusted JSON and return typed nodes or raise ValidationError."""
    return CONTENT_ADAPTER.validate_python(value)


def content_json_schema() -> dict[str, object]:
    """Return the canonical schema used by frontend/editor integrations."""
    return CONTENT_ADAPTER.json_schema(by_alias=True)


if __name__ == "__main__":
    print(json.dumps(content_json_schema(), indent=2, ensure_ascii=False))
